### DISCUSSION TITLE
how to use fst howtwords in python pipline?

### DISCUSSION THREAD

**Discussion by zyb8543d** [modelscope/FunASR#1085] [Q&A] [answered]
c++ decoding already supports hot words，how to use fst howtwords in python pipline?

**Comment by LauraGPT** [comment]
Sorry, we do not have this plan.

**Comment by LauraGPT** [comment, marked as answer]
Update for current FunASR: the 2023 reply predates the current Python hotword API.

There is one important distinction:

- The local Python `AutoModel` pipeline does **not** load a compiled WFST/FST graph.
- `paraformer-zh` supports SeACo neural hotword bias through the singular `hotword` argument.
- Weighted WFST/FST hotwords remain a C++/ONNX runtime feature.

For the current Python pipeline:

```python
from funasr import AutoModel

model = AutoModel(
    model="paraformer-zh",
    vad_model="fsmn-vad",
    punc_model="ct-punc",
    device="cpu",
    disable_update=True,
)

wav_file = f"{model.model_path}/example/asr_example.wav"

plain = model.generate(
    input=wav_file,
    batch_size_s=300,
)[0]

biased = model.generate(
    input=wav_file,
    batch_size_s=300,
    hotword="魔搭",
)[0]

print(plain["text"])
print(biased["text"])
```

Verified with FunASR 1.3.14 on CPU:

```text
欢迎大家来到么哒社区进行体验。
欢迎大家来到魔搭社区进行体验。
```

Both results contained the same 14 timestamps; only the biased term changed.

For multiple hotwords, use a whitespace-separated string:

```python
model.generate(input=wav_file, hotword="达摩院 魔搭")
```

You can also pass a UTF-8 `.txt` file path, with one hotword per line:

```text
达摩院
魔搭
```

```python
model.generate(input=wav_file, hotword="/path/to/hotwords.txt")
```

The current parser for string, local-file, and URL inputs is here: [`generate_hotwords_list`](https://github.com/modelscope/FunASR/blob/1d8080af99fad370645eec7de9d1bcc5ae4ab587/funasr/models/seaco_paraformer/model.py#L583-L690). The same `paraformer-zh` usage is in the [Python tutorial](https://github.com/modelscope/FunASR/blob/1d8080af99fad370645eec7de9d1bcc5ae4ab587/docs/tutorial/README.md#speech-recognition-non-streaming).

For **Fun-ASR-Nano**, the API is different and uses the plural list argument:

```python
model.generate(input=wav_file, hotwords=["达摩院", "魔搭"])
```

If you specifically require weighted WFST/FST decoding, use the C++/ONNX runtime or server and pass a weighted hotword file with `--hotword`:

```text
阿里巴巴 20
魔搭 15
```

That path is documented in the [online runtime guide](https://github.com/modelscope/FunASR/blob/1d8080af99fad370645eec7de9d1bcc5ae4ab587/runtime/docs/SDK_advanced_guide_online.md#L90-L126) and [offline runtime guide](https://github.com/modelscope/FunASR/blob/1d8080af99fad370645eec7de9d1bcc5ae4ab587/runtime/docs/SDK_advanced_guide_offline.md#L57-L68). A Python websocket client may call that runtime server, but that is different from loading an FST into the local Python `AutoModel` pipeline.

For very large vocabularies, the tutorial also documents `postprocess_hotwords`; that is text-level correction after recognition, not decoder or WFST bias.
